plugin/output/s3/s3.go

```diff
@@ -262,3 +290,122 @@ func (p *Plugin) generateObjectName(name string) string {
 	objectName = objectName[0 : len(objectName)-len(p.compressor.getExtension())]
 	return fmt.Sprintf("%s.%s%s", objectName, n, p.compressor.getExtension())
 }
+
+func (p *Plugin) createMinioClients() (minioClients map[string]objectStoreClient) {
+	minioClients = make(map[string]objectStoreClient)
+	// initialize minio clients object for main bucket.
+	mainClient, err := minio.New(p.config.Endpoint, p.config.AccessKey, p.config.SecretKey, p.config.Secure)
+	if err != nil || mainClient == nil {
+		p.logger.Panicf("could not create minio client, error: %s", err.Error())
+	}
+	minioClients[p.config.Bucket] = mainClient
+
+	// TODO create minioClients on upper level, dependency inversion will be fixed, mock will be set in test, not in Config.client.
+	if p.config.client != nil {
+		p.logger.Info("set mock clients")
+		minioClients[p.config.Bucket] = *p.config.client
+	}
+
+	for _, singleB := range p.config.MultiBuckets {
+		bClient, err := minio.New(singleB.Endpoint, singleB.AccessKey, singleB.SecretKey, singleB.Secure)
+		if err != nil || mainClient == nil {
+			p.logger.Panicf("could not create minio client, error: %s", err.Error())
+		}
+		minioClients[singleB.Bucket] = bClient
+
+		// TODO create minioClients on upper level, dependency inversion will be fixed, mock will be set in test, not in Config.client.
+		if p.config.client != nil {
+			p.logger.Info("set mock clients")
+			minioClients[singleB.Bucket] = *p.config.client
+		}
+	}
+
+	return minioClients
+}
+
+func (p *Plugin) setDirs(outPlugCount int) {
+	// dir for all bucket files.
+	dir, _ := filepath.Split(p.config.FileConfig.TargetFile)
+
+	targetDirs := make(map[string]string, outPlugCount)
+	targetDirs[p.config.Bucket] = dir
+	for _, singleB := range p.config.MultiBuckets {
+		targetDirs[singleB.Bucket] = fmt.Sprintf("%s%s/", dir, singleB.Bucket)
+	}
+	p.targetDirs = targetDirs
+}
+
+func (p *Plugin) setFileNames(outPlugCount int) {
+	_, f := filepath.Split(p.config.FileConfig.TargetFile)
+	p.fileExtension = filepath.Ext(f)
+	fileNames := make(map[string]string, outPlugCount)
+
+	mainFileName := f[0 : len(f)-len(p.fileExtension)]
+	fileNames[p.config.Bucket] = mainFileName
+	for _, singleB := range p.config.MultiBuckets {
+		fileNames[singleB.Bucket] = singleB.Bucket
+	}
+	p.fileNames = fileNames
+}
+
+func (p *Plugin) startPlugins(Params *pipeline.OutputPluginParams, outPlugCount int) {
+	outPlugins := make(map[string]*file.Plugin, outPlugCount)
+	// Checks required bucket and exit if it's invalid.
+	exist, err := p.clients[p.config.Bucket].BucketExists(p.config.Bucket)
+	if err != nil {
+		p.logger.Panicf("could not check bucket: %s, error: %s", p.config.Bucket, err.Error())
+	}
+	if !exist {
+		p.logger.Fatalf("bucket: %s, does not exist", p.config.Bucket)
+	}
+	p.logger.Infof("bucket %s exists", p.config.Bucket)
+
+	anyPlugin, _ := file.Factory()
```

Comment:
may decrease of duplicate code here. Here is example:

```
func (p *Plugin) createPlugin(bucket string) (*file.Plugin, error) {
	exist, err := p.clients[bucket].BucketExists(bucket)
	if err != nil {
		return nil, err
	}
	if !exist {
		return nil, errors.New("todo")
	}
	p.logger.Infof("bucket %s exists", bucket)

	anyPlugin, _ := file.Factory()
	outPlugin := anyPlugin.(*file.Plugin)
	outPlugin.SealUpCallback = p.addFileJobWithBucket(bucket)
	return outPlugin, nil
}
```

Reply:
DRY!